### ISSUE TITLE
Incorrect PDF page numbering with added front- and back-matter

### ISSUE THREAD

**Issue by gwyndaf** [ChordPro/chordpro#269]
Minor issue. I really like the use of different PDF page number format for added for front matter (small Roman) and back matter (back-n). That is, the numbering shown in PDF reader navigation (e.g. qpdfview or xreader on Linux), not what's shown on the page.

However, they don't always seem accurate.

__Real-life example:__ 3-page front matter, 133-page songbook body (incl. 3-page ToC), 3-page back matter
[biscuit_bundle_TEST_(Guitar).pdf](https://github.com/ChordPro/chordpro/files/10888119/biscuit_bundle_TEST_.Guitar.pdf)
a) Correct front matter numbering (i-iii).
b) Wrong number of songbook body pages (only 130, not 133 with Arabic numbering, 3 pages too few).
c) Wrong back matter numbering, counts 4 pages (not 3).
d) 2 final output pages have no numbering: maybe a consequence of b) and c).

__Test sample:__ 2-page front matter, 3 page songbook body (incl. 1-page ToC), 4-page back matter (incl. blank)
[TestBook_custom.pdf](https://github.com/ChordPro/chordpro/files/10888174/TestBook_custom.pdf)
a) Incorrect front matter numbering (page i only - expected i-ii)
b) Wrong number of songbook body pages (only 2, not 3 with Arabic numbering)
c) Wrong back matter numbering, counts 5 pages (not 4)
d) 1 final output page has no numbering (reflects a), b), c) above)

I'm not exactly sure how to make best use of this evidence, but a few observations:

* The range of Roman numbering given to the front matter equals the number of ToC pages. I vaguely remember an earlier version using lower-case Roman numbering for the ToC (not front matter), so maybe this is a carry-over?
* Body page numbering stops short by a number equal to the ToC size, so maybe the same cause again?
* Number of back matter pages seems to be 1 greater than the actual number of pages.
* Together, the above factors account for the shortfall in numbering at the end.
* My config has `"even-odd-pages" : 0` and `"pagealign-songs" : 0`. I haven't tried to complicate the issue by enabling odd/even pages!
* Processing the test (but not production) sample gives an message: `Use of uninitialized value $dat in string at /usr/local/share/perl/5.36.0/PDF/API2/Basic/PDF/Filter/FlateDecode.pm line 42.` but outputs OK, so that might not be relevant.

Are there any other tests I could run that might give further insight?

Currently running release v6.000 on Debian Bookworm but, looking back at my archive, it's not a new issue. However, it's clearly so minor that I didn't notice it before!

**Comment by sciurius** [comment]
I'm afraid I cannot quite follow...
Given Testbook-custom.pdf this is what I see in the viewer:

![scrot20230304212516](https://user-images.githubusercontent.com/1118804/222927250-dcb8df20-7b2e-4bd3-923e-4ef9ade3572d.png)

So there is two front-matter pages `front-1` and `front-2`. A ToC `i`, two song pages `1` and `2` and the four back-matter pages `back-1` through `back-4`. Can you point out what is wrong?

**Comment by gwyndaf** [comment]
Oh, that's interesting. It might be a PDF reader issue then.

I get the numbering below using xreader, but will try some others to see if I can replicate the issue across other PDF readers. I get the same numbering as below with qpdfview, but think that uses the same Poppler PDF library, so that might be the root of the issue.

![TestBook - xreader screenshot](https://user-images.githubusercontent.com/1415754/222928091-3d4e2ab1-af6f-4978-9a88-af6a49f8cd5c.png)

**Comment by sciurius** [comment]
![Uploading scrot20230304222848.png…]()

**Comment by gwyndaf** [comment]
Oh, this is strange. I have the same Xreader version 3.6.3 but apparently different results, so I'll need to investigate further.

Meanwhile, I've discovered that I should have used the terminology _page labels_ to differentiate from sequential page numbers.

**Comment by gwyndaf** [comment]
Some interesting progress, by extracting page labels. TLDR: it's probably not a ChordPro issue.

`pdftk dump_data` gives me this, which seems to match the expected experience you show above:
```
PageLabelBegin
PageLabelNewIndex: 0
PageLabelStart: 1
PageLabelPrefix: front-
PageLabelNumStyle: DecimalArabicNumerals    << so front-1 to front-2
PageLabelBegin
PageLabelNewIndex: 2
PageLabelStart: 1
PageLabelNumStyle: LowercaseRomanNumerals    << so i
PageLabelBegin
PageLabelNewIndex: 3
PageLabelStart: 1
PageLabelNumStyle: DecimalArabicNumerals    << so 1 to 2
PageLabelBegin
PageLabelNewIndex: 5
PageLabelStart: 1
PageLabelPrefix: back-
PageLabelNumStyle: DecimalArabicNumerals    << so back-1, back-2 until ?end
```

[This](https://github.com/HeimMatthias/pdfpagelabels) program to dump labels (based on Poppler, I believe) gives:
```
i
1
2
back-1
back-2
back-3
back-4
back-5
```
This exactly matches my experience, including only 8 labels, so last page has none.

I _think_ this suggests that the internal PDF structure is as expected, but gets interpreted strangely by some readers, with Poppler or my systems as likely suspects. More testing needed to narrow down the exact cause...

**Comment by sciurius** [comment]
> I think this suggests that the internal PDF structure is as expected

Yes, verified that with [rups](https://itextpdf.com/products/rups).

The poppler version I have here is 22.08.0.

**Comment by gwyndaf** [comment]
My Poppler is v22.12.0, so maybe it's become less tolerant! I didn't know about RUPS, but it seem a useful tool for examining PDFs.

Testing across a variety of PDF readers gave similar results to xreader. SumatraPDF on Windows (which I believe uses muPDF rendering) had page labelling the same as my earlier example, while Acrobat DC interestingly didn't show labels at all and reverted to sequential page numbers.

So I created the same page label structure in a dummy/blank InDesign document and exported that to PDF, which behaves as expected in a variety of readers and _does_ display page labels (not numbers) in Acrobat. I don't know the detailed PDF standard, but reckoned that Adobe product should export in a compliant way.
[PageLabel_test.pdf](https://github.com/ChordPro/chordpro/files/10891958/PageLabel_test.pdf)

The same pdftk dump_data of that PDF (from InDesign) gives:
```
PageLabelBegin
PageLabelNewIndex: 1
PageLabelStart: 1
PageLabelPrefix: front-
PageLabelNumStyle: DecimalArabicNumerals
PageLabelBegin
PageLabelNewIndex: 3
PageLabelStart: 1
PageLabelNumStyle: LowercaseRomanNumerals
PageLabelBegin
PageLabelNewIndex: 4
PageLabelStart: 1
PageLabelNumStyle: DecimalArabicNumerals
PageLabelBegin
PageLabelNewIndex: 6
PageLabelStart: 1
PageLabelPrefix: back-
PageLabelNumStyle: DecimalArabicNumerals
```
This looks _almost_ the same as my previous example, but the interesting difference is that the first `PageLabelNewIndex` value is `1`, not `0`, and subsequent values are 1 greater than my ChordPro output. This suggests that Adobe's creation of PDFs begins with a page number/index of 1, but the ChordPro output (via PDF::API2 maybe) begins with 0.

I'll examine both with RUPS once it's compiled...

**Comment by sciurius** [comment]
When inspecting with RUPS I see one difference: ChordPro designates the start as -1, while Adobe uses 0.
That could make a difference for programs that take it more precise.

So maybe ChordPro is wrong after all, and atril and poppler 22.08.0 are too forgiving...

**Comment by sciurius** [comment]
Nailed it. It is a subtle difference between PDF::API2 2.042 and older versions. Before 2.042, the pageLabels method took a 0-based page number, but the new 2.042 method page_labels requires a 1-based page number.

Assuming you have PDF::API2 2.042 or later, you can change `lib/App/Music/ChordPro/Output/PDF/Writer.pm` around line 439 by adding `+1` after `$page`:

````
	$c->( $self->{pdf}, $page+1, %$opts );
````

**Comment by gwyndaf** [comment]
Ah, that's great.

I suspected maybe there was some change in PDF standards and/or how strictly they're applied.

**Comment by sciurius** [comment]
Fixed in 6.000_010.

**Comment by gwyndaf** [comment]
That's fantastic. Now works as expected with test and production samples, and RUPS examination :)
